### DISCUSSION TITLE
my_uart_putc callback function example

### DISCUSSION THREAD

**Discussion by andkae** [charlesnicholson/nanoprintf#240] [General]
Hi Charles,

In the Readme is the usage of ```npf_pprintf``` demonstrated.
```c
void print_to_uart(void) {
  npf_pprintf(&my_uart_putc, NULL, "Hello %s%c %d %u %f\n", "worl", 'd', 1, 2, 3.f);
}
```

At the moment I stumble a little to write the callback function ```my_uart_putc```. Could you please share the prototype for such an callback function.

Thanks in advance.

BR,
Andreas

**Comment by mickjc750** [comment]
The function pointer is typedef'd on line 38 of the header:
typedef void (*npf_putc)(int c, void *ctx);

So the prototype in your case would be:
void my_uart_putc(int c, void* ctx);

And your function might look something like:
void my uart_putc(int c, void* ctx)
{
 (void)ctx;  // this can avoid a compilation warning if ctx is unused
 UartWrite((char)c);
}

**Reply by andkae** [reply]
Thanks :) I'll try.

**Comment by charlesnicholson** [comment]
Thanks for chiming in with the correct answer, @mickjc750 :)

I'll file an issue to improve the documentation, @andkae's question makes it clear that this is a gap.

I filed #241 to address it, sorry it wasn't clear enough! (I take these things for granted, as the author :))

**Reply by mickjc750** [reply]
It may also be worth explaining how to wrap the non-variadic npf_vpprintf() with a variadic uart_printf() or lcd_printf().
 printf tends to be called a lot where it's used, and a wrapper here would be more code size efficient than passing the character handler and a NULL for every call.
 stdio provides printf as a way of formatting text to file IO, as it's a file IO interface. So it makes sense (to me anyway) that a uart module or LCD module should also provide text formatting to their respective device. This is something I've explained in my prnf repository, which seems to be less popular, possibly due to it deviating from the standard printf behaviour, specifically on %n.

**Reply by charlesnicholson** [reply]
@mickjc750 Do you mean something along the lines of this: https://github.com/charlesnicholson/nanoprintf/blob/main/examples/wrap_npf/your_project_printf.cc

But calling into `npf_vpprintf` instead of `npf_vsnprintf`? That's certainly easy enough; I took for granted that the existing example would be sufficient to show the style of usage that you're talking about as well...

There's certainly no harm in more examples, though.
